SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LoadCopyToBorrowOptimization.swift

```diff
@@ -0,0 +1,264 @@
+//===--- LoadCopyToBorrowOptimization.swift --------------------------------==//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Replaces a `load [copy]` with a `load_borrow` if possible.
+///
+/// ```
+///   %1 = load [copy] %0
+///   // no writes to %0
+///   destroy_value %1
+/// ```
+/// ->
+/// ```
+///   %1 = load_borrow %0
+///   // no writes to %0
+///   end_borrow %1
+/// ```
+///
+/// The optimization can be done if:
+/// * During the (forward-extended) lifetime of the loaded value the memory location is not changed.
+/// * All (forward-extended) uses of the loaded value support guaranteed ownership.
+/// * The (forward-extended) lifetime of the loaded value ends with `destroy_value`(s).
+///
+let loadCopyToBorrowOptimization = FunctionPass(name: "load-copy-to-borrow-optimization") {
+  (function: Function, context: FunctionPassContext) in
+
+  if !function.hasOwnership {
+    return
+  }
+
+  for inst in function.instructions {
+    if let load = inst as? LoadInst {
+      optimize(load: load, context)
+    }
+  }
+}
+
+private func optimize(load: LoadInst, _ context: FunctionPassContext) {
+  if load.loadOwnership != .copy {
+    return
+  }
+
+  var ownershipUses = OwnershipUses(context)
+  defer { ownershipUses.deinitialize() }
+
+  // Handle the special case if the `load` is immediately followed by a single `move_value`.
+  // Note that `move_value` is _not_ a forwarding instruction.
+  // In this case we have to preserve the move's flags by inserting a `begin_borrow` with the same flags.
+  // For example:
+  //
+  //   %1 = load [copy] %0
+  //   %2 = move_value [lexical] %1
+  //    ...
+  //   destroy_value %2
+  // ->
+  //    %1 = load_borrow %0
+  //    %2 = begin_borrow [lexical] %1
+  //     ...
+  //    end_borrow %2
+  //    end_borrow %1
+  //
+  let loadedValue = load.singleMoveValueUser ?? load
+
+  // This is required if the lifetime of the `load` ends up in a dead-end block and the `destroy_value` is
+  // missing at the `unreachable`. In such a case the dataflow would _not_ visit potential writes to the load's
+  // memory location. In the following example, the `load [copy]` must not be converted to a `load_borrow`:
+  //
+  //   %1 = load [copy] %0
+  //     ...
+  //   store %2 to %0
+  //     ...          // no destroy of %1!
+  //   unreachable    // the lifetime of %1 ends here
+  //
+  context.completeLifetime(of: loadedValue)
```

Comment:
Generally, lifetimes need to be completed inside out/bottom up (as in `SILGenCleanup::completeOSSALifetimes`).  I haven't come up with an example where that would be a problem in this usage, but I'm not confident it's not.